---
title: "Re-analysing Patchett 2020 data"
author: "Anne-Lise GÃ©rard"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 5
    fig_height: 5
theme: cosmo
---

Source codes: https://github.com/Anne-LiseGerard/dftd_rnaseq
Data: https://doi.org/10.1007/s00018-019-03259-2

```{r,pkg}

suppressPackageStartupMessages({
    library("reshape2")
    library("gplots")
    library("DESeq2")
    library("mitch")
    library("limma")
    library("kableExtra")
    library("dplyr")
    library("tidyr")
    library("ComplexHeatmap")
    library("RColorBrewer")
    library("circlize")
    library("pathview")
    library("stringr")
    library("readxl")
    library("grid")
    library("ggplot2")
    library("viridis")
    library("enrichplot")
    library("SBGNview")
})

knitr::opts_chunk$set(dev = 'svg')

```

## Background

Goal: re-analyse previously published DFTD tumour biopsy transcriptomes and compare to DFTD cell line transcriptomes, focusing on genes involved in metabolism. Below is a list of the samples for this project.

```{r, ss}

ss <- read.table("../ss_patchett.txt",sep="\t",fill=TRUE,header=TRUE)
ss$DFT <- as.factor(ss$DFT)

ss %>%
  kbl(caption="Sample sheet for all samples") %>%
  kable_paper("hover", full_width = F)

```

## Functions

```{r, func}

maplot <- function(de,contrast_name) {
  sig <-subset(de, padj < 0.05 )
  up <-rownames(subset(de, padj < 0.05 & log2FoldChange > 0))
  dn <-rownames(subset(de, padj < 0.05 & log2FoldChange < 0))
  GENESUP <- length(up)
  GENESDN <- length(dn)
  DET=nrow(de)
  SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down", DET, "detected")
  ns <-subset(de, padj > 0.05 )
  plot(log2(de$baseMean),de$log2FoldChange,
       xlab="log2 basemean", ylab="log2 foldchange",
       pch=19, cex=1, col="dark gray",
       main=contrast_name, cex.main=0.7)
  points(log2(sig$baseMean),sig$log2FoldChange,
         pch=19, cex=1, col="#5297A7")
  mtext(SUBHEADER,cex = 0.7)
  grid()
}

make_volcano <- function(de,name) {
    sig <- subset(de,padj<0.05)
    N_SIG=nrow(sig)
    N_UP=nrow(subset(sig,log2FoldChange>0))
    N_DN=nrow(subset(sig,log2FoldChange<0))
    DET=nrow(de)
    HEADER=paste(N_SIG,"@5%FDR,", N_UP, "up", N_DN, "dn", DET, "detected")
    plot(de$log2FoldChange,-log10(de$padj),cex=1,pch=19,col="#D5D7E2",
        main=name, xlab="log2 FC", ylab="-log10 pval")
    mtext(HEADER)
    grid()
    points(sig$log2FoldChange,-log10(sig$padj),cex=1,pch=19,col="#5297A7")
}

custom_heatmap <- function(zscore, fc, aveexpr, title) {
  h1 <- Heatmap(zscore, cluster_rows = F,
              column_labels = colnames(zscore), name="Z-score",
              cluster_columns = T,
              col= colorRamp2(c(min(zscore),0,max(zscore)), hcl_palette = "Blue-Red 2"))
  h2 <- Heatmap(fc, row_labels = rownames(fc), row_names_gp = gpar(fontsize = 6),
              cluster_rows = F, name="logFC", top_annotation = ha, col = col_logFC,
              cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
                grid.text(round(as.numeric(fc[i, j],2)), x, y,gp=gpar(fontsize=6))
              }, column_title = title)
  # h3 <- Heatmap(aveexpr, row_labels = rownames(aveexpr), row_names_gp = gpar(fontsize = 6),
  #             cluster_rows = F, name = "AveExpr", col=col_AveExpr,
  #             cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
  #               grid.text(round(as.numeric(aveexpr[i, j],2)), x, y,gp=gpar(fontsize=6))},
  #             column_title = title)

  h <- h1 + h2 #+ h3
  h
}

```

## Load data

Import data from the aligner. 

```{r, import}

tmp <- read.table("../fastq/3col_patchett.tsv.gz")
tmp$V1 <- gsub('.fastq-trimmed.fastq','',tmp$V1)
x <- as.data.frame(acast(tmp, V2~V1, value.var="V3",fun.aggregate=sum))

tmp_old <- read.table("../fastq/3col_old.tsv.gz")
tmp_old$V1 <- gsub('.fastq-trimmed.fastq','',tmp_old$V1)
x_old <- as.data.frame(acast(tmp_old, V2~V1, value.var="V3",fun.aggregate=sum))

```

Load gene names.

```{r, genenames}

gn <- read.table("../ref/Sarcophilus_harrisii.mSarHar1.11.cdna+ncrna.genenames.tsv",fill=TRUE)
gn <- gn[order(gn$V1),]

gn_old <- read.table("../ref/Sarcophilus_harrisii.DEVIL7.0.cdna+ncrna.genenames.tsv",fill=TRUE)
gn_old <- gn_old[order(gn_old$V1),]

```

```{r, homology}

hm <- read.table("../ref/mart_export_ensembl109_2023-07-14.txt",sep="\t",header=TRUE)

hm_old <- read.table("../ref/mart_export_ensembl101_2023-09-06.txt",sep="\t",header=TRUE)

```

Now need to collapse transcript data to genes.

```{r, collapse}

x$gene <- paste(gn$V2,gn$V3)

y <- aggregate(. ~ gene,x,sum)

rownames(y) <- y$gene
y$gene = NULL


x_old$gene <- paste(gn_old$V2,gn_old$V3)

y_old <- aggregate(. ~ gene,x_old,sum)

rownames(y_old) <- y_old$gene
y_old$gene = NULL

```

## Quality control

Samples with <1M reads should be omitted.
Will also round values to integers.
Using the new genome yields a higher number of mapped reads.

```{r, qc1}

cs <- colSums(y)
cs <- cs[order(cs)]

par(mar=c(5,10,5,2))
barplot(cs,main="All samples",horiz=TRUE,las=1)
abline(v=1e7,col="red",lty=2)

y <- round(y)


cs_old <- colSums(y_old)
cs_old <- cs_old[order(cs_old)]

par(mar=c(5,10,5,2))
barplot(cs_old,main="All samples - old genome ",horiz=TRUE,las=1, xli=c(0,4e+07))
abline(v=1e7,col="red",lty=2)

y_old <- round(y_old)

```

## MDS

This will help us to visualise the sources of variability in the overall dataset.

Plot MDS. Also fix the sample names.

Plot MDS by DFT.

```{r, plotmds1}

par(mar = c(5.1, 4.1, 4.1, 2.1))

colnames(y) <- ss$sample_id

saveRDS(y, file = "y_biopsies.rds")

cs <- colSums(y)
cs <- cs[order(cs)]

par(mar=c(5,10,5,2))
barplot(cs,main="All samples",horiz=TRUE,las=1)


par(mar = c(5.1, 4.1, 4.1, 2.1))

colnames(y_old) <- ss$sample_id[1:8]

cs_old <- colSums(y_old)
cs_old <- cs_old[order(cs_old)]

par(mar=c(5,10,5,2))
barplot(cs_old,main="All samples - old genome",horiz=TRUE,las=1, xlim=c(0,4e+07))


par(mar = c(5.1, 4.1, 4.1, 2.1) )

cols <- ss$DFT
cols <- gsub("DFT1","#B3B9DF",cols)
cols <- gsub("DFT2","#DDAFB7",cols)
mymds <- plotMDS(y,plot=FALSE)

mymds_old <- plotMDS(y_old,plot=FALSE)

# fix the xlims
XMIN=min(mymds$x)
XMAX=max(mymds$x)
XMID=(XMAX+XMIN)/2
XMIN <- XMID + (XMIN-XMID)*1.1
XMAX <- XMID+(XMAX-XMID)*1.1
plotMDS(mymds,pch=19,cex=3,col=cols,main="MDS plot",xlim=c(XMIN,XMAX))
text(mymds,labels=colnames(y))
mtext("blue=DFT1, pink=DFT2")

plotMDS(mymds_old,pch=19,cex=3,col=cols,main="MDS plot - old genome",xlim=c(XMIN,XMAX))
text(mymds_old,labels=colnames(y_old))
mtext("blue=DFT1, pink=DFT2")

y_DFT1 <- y[,colnames(y) %in% c("sha_DFT1_1-1","sha_DFT1_1-2","sha_DFT1_2-1","sha_DFT1_2-2",
                                "C5065_UT_01", "C5065_UT_02")]
y_DFT2 <- y[,colnames(y) %in% c("sha_DFT2_1-1","sha_DFT2_1-2","sha_DFT2_2-1","sha_DFT2_2-2",
                                "sha_DFT2_RV_1-1","sha_DFT2_RV_1-2","sha_DFT2_RV_2-1","sha_DFT2_RV_2-2")] 

y_biopsy <- y[,colnames(y) %in% c("sha_DFT1_1-1","sha_DFT1_1-2","sha_DFT1_2-1","sha_DFT1_2-2",
                                  "sha_DFT2_1-1","sha_DFT2_1-2","sha_DFT2_2-1","sha_DFT2_2-2")]
y_cline <- y[,colnames(y) %in% c( "C5065_UT_01", "C5065_UT_02",
                                "sha_DFT2_RV_1-1","sha_DFT2_RV_1-2","sha_DFT2_RV_2-1","sha_DFT2_RV_2-2")] 

mdsDFT1 <- plotMDS(y_DFT1,plot=FALSE)
plotMDS(mdsDFT1,pch=19,cex=3,col="#B3B9DF",main="MDS plot",xlim=c(XMIN,XMAX))
text(mdsDFT1,labels=colnames(y_DFT1))
mtext("DFT1")

mdsDFT2 <- plotMDS(y_DFT2,plot=FALSE)
plotMDS(mdsDFT2,pch=19,cex=3,col="#DDAFB7",main="MDS plot",xlim=c(XMIN,XMAX))
text(mdsDFT2, labels=colnames(y_DFT2))
mtext("DFT2")

mds_biopsy <- plotMDS(y_biopsy,plot=FALSE)
plotMDS(mds_biopsy,pch=19,cex=3,col=cols,main="Tumour biopsies",xlim=c(XMIN,XMAX))
text(mds_biopsy,labels=colnames(y_biopsy))
mtext("Tumour biopsies: blue=DFT1, pink=DFT2")

mds_cline <- plotMDS(y_cline,plot=FALSE)
plotMDS(mds_cline,pch=19,cex=3,col=cols,main="MDS plot",xlim=c(XMIN,XMAX))
text(mds_cline,labels=colnames(y_cline))
mtext("Cell lines: blue=DFT1, pink=DFT2")

```

## DESeq2

Sum replicates then run differential expression analysis.
For now, ignore alignment on old genome and cell line data.

```{r, de}

y_cell <- y[,9:14]
y <- y[,1:8]

# combine replicates for each sample
DFT1_1 <- rowSums(y[,ss$sample=="DFT1_1"])
DFT1_2 <- rowSums(y[,ss$sample=="DFT1_2"])
DFT2_1 <- rowSums(y[,ss$sample=="DFT2_1"])
DFT2_2 <- rowSums(y[,ss$sample=="DFT2_2"])

y2 <- data.frame(DFT1_1,DFT1_2,DFT2_1,DFT2_2)
ss2 <- as.data.frame(colnames(y2))
colnames(ss2) <- "sample"
ss2$DFT <- factor(c("DFT1","DFT1","DFT2","DFT2"))

y2 <- y2[which(rowMeans(y2)>10),] # remove genes with average count < 10
dim(y2)

dds <- DESeqDataSetFromMatrix(countData = y2 , colData = ss2, design = ~ DFT )
dds2 <- dds
res <- DESeq(dds)
z <- results(res)
vsd <- vst(dds)
zz<-cbind(as.data.frame(z),assay(vsd))
dge<-as.data.frame(zz[order(zz$pvalue),])
dge2 <- dge
sig <- subset(dge,padj<0.05)
sig2 <- sig
sig2_up <- rownames(subset(sig,log2FoldChange>0))
sig2_dn <- rownames(subset(sig,log2FoldChange<0))
length(sig2_up)
length(sig2_dn)

```

```{r,deviz}

maplot(dge2,"Tumour biopsies")

make_volcano(dge2,"Tumour biopsies")

sig[1:50,1:6] %>%
  kbl(caption="Tumour biopsies") %>%
  kable_paper("hover", full_width = F)

write.table(dge2,file="dge3.tsv",sep="\t")


```

```{r,de2vizheatmap,fig.dim = c(4, 6)}

# heatmap
mx <- sig[,7:ncol(sig)]
mx <- head(mx,30)
colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(as.matrix(mx),trace="none",scale="row",
  col=colfunc(25),ColSideColors=cols[3:6],mar=c(8,14),
  cexRow=0.7,cexCol=1)

# top 20 up & top 20 down
sig <- sig[order(sig$log2FoldChange, decreasing=TRUE),]

sig_noNA <- sig
sig_noNA$genes <- rownames(sig_noNA)
sig_noNA <- filter(sig_noNA, !grepl(' NA', genes))

top <- rbind(head(sig_noNA,20),tail(sig_noNA,20))

mx <- top[,7:ncol(top)] # get normalised counts

mx.scaled <- t(apply(mx[,1:4], 1, scale)) # center and scale each column (Z-score)
colnames(mx.scaled) <- colnames(mx[,1:4])
colnames(mx.scaled) <- c("DFT1 1", "DFT1 2", "DFT2 1", "DFT2 2")
log2FC <- as.matrix(top$log2FoldChange)
rownames(log2FC) <- rownames(top)
colnames(log2FC) <- "logFC"
mean <- as.matrix(top$baseMean)
rownames(mean) <- rownames(top)
colnames(mean) <- "AveExpr"


col_logFC <- col_logFC <- colorRamp2(c(-20,0,20), hcl_palette = "Blue-Red 2")
#col_AveExpr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white","red"))

ha <- HeatmapAnnotation(summary=anno_summary(gp=gpar(fill="#D5D7E2"),height=unit(2, "cm")))

custom_heatmap(mx.scaled, log2FC, mean, title = "Top 40 DEGs - Tumour biopsies")

```
## Enrichment analysis

Need to get human homologs of these genes.
These were obtained from Ensembl v109 biomart (website).

```{r}

rownames(dge2) <- sapply(strsplit(rownames(dge2),"\\."),"[[",1)

hm2 <- hm[hm$Tasmanian.devil.gene.stable.ID != "",]

gt <- hm2[,2:3]
length(unique(gt$Tasmanian.devil.gene.stable.ID))
length(unique(gt$Gene.name))

for(i in 1:length(gt$Gene.name)){
  if(gt$Gene.name[i]==""){gt$Gene.name[i] <- gt$Tasmanian.devil.gene.stable.ID[i]}
}

saveRDS(gt, "~/dftd_RNAseq_annelise/dge/gt.rds")

```

Now run mitch for DGE2. The pathways are sourced from Reactome 7th July 2023.

```{r,mitch1, fig.dim = c(12, 22)}

genesets <- gmt_import("../ref/ReactomePathways_2023-07-14.gmt")

m2 <- mitch_import(dge2, DEtype="deseq2", geneTable=gt)
head(m2)

res2 <- mitch_calc(m2, genesets, priority="effect",cores=16)

if ( !file.exists("mitch_biopsies_reactome.html") ) {
  mitch_report(res2, "mitch_biopsies_reactome.html")
}

top <- subset(res2$enrichment_result,p.adjustANOVA<0.05)
saveRDS(top, "~/dftd_RNAseq_annelise/dge/gsea_biopsies.rds")

top %>%
  kbl(caption="Enriched pathways") %>%
  kable_paper("hover", full_width = F)

ggplot(top, aes(x = s.dist, y = reorder(set, s.dist), color = p.adjustANOVA, size = setSize)) + 
  geom_point(stat = 'identity') + 
  scale_color_viridis(option="mako", direction=1,limits = c(0, 0.05)) +
  xlab("s dist") + ylab("pathway") + ggtitle("Reactome ALL") + 
  theme_bw()

```
```{r,mitch2, fig.dim = c(8, 6)}
top20 <- rbind(slice_max(top, order_by=s.dist, n=10), slice_min(top, order_by=s.dist, n=10))

ggplot(top20, aes(x = s.dist, y = reorder(set, s.dist), color = p.adjustANOVA, size = setSize)) + 
  geom_point(stat = 'identity') + 
  scale_color_viridis(option="mako", direction=1,limits = c(0, 0.05)) +
  xlab("s dist") + ylab("pathway") + ggtitle("Reactome top 20") + 
  theme_bw()
```
```{r,mitch3, fig.dim = c(12, 6)}
react_metabosets <- readRDS("../ref/reactome_metabo.rds")

metabo1 <- filter(top, set %in% names(react_metabosets))
saveRDS(metabo1, "~/dftd_RNAseq_annelise/dge/gsea_biopsies_metabo.rds")

ggplot(metabo1 , aes(x = s.dist, y = reorder(set, s.dist), color = p.adjustANOVA, size = setSize)) + 
  geom_point(stat = 'identity') + 
  scale_color_viridis(option="mako", direction=1,limits = c(0, 0.05)) +
  xlab("s dist") + ylab("pathway") + ggtitle("Reactome metabolism") + 
  theme_bw()

```

KEGG gene sets were sourced from MSigDb on 14 September 2023.

```{r, mitch4}

genesets2 <- gmt_import("../ref/c2.cp.kegg.v2023.1.Hs.symbols.gmt.txt")

res2 <- mitch_calc(m2, genesets2, priority="effect",cores=16)

if ( !file.exists("mitch_biopsies_kegg.html") ) {
  mitch_report(res2, "mitch_biopsies_kegg.html")
}

top <- subset(res2$enrichment_result,p.adjustANOVA<0.05)

topup <- head(subset(top,s.dist>0),20)

topup %>%
  kbl(caption="Top 20 upregulated pathways") %>%
  kable_paper("hover", full_width = F)

topdn <- head(subset(top,s.dist<0),20)

topdn %>%
  kbl(caption="Top 20 downregulated pathways") %>%
  kable_paper("hover", full_width = F)

top2 <- rbind(head(topup,10),head(topdn,10))
top2 <- top2[order(-top2$s.dist),]
par(mar=c(5,15,5,2))

barplot(rev(abs(top2$s.dist)),horiz=TRUE,las=1,names.arg=rev(top2$set),
  main="Top KEGG Pathways",cex.names=0.6,xlab="Enrichment Score",
  col=c(rep("blue",nrow(head(topdn,10))),rep("red",nrow(head(topup,10)))),
  xlim=c(0,0.8))

kegg_metabosets <- gmt_import("../ref/gsea/gsea_KEGG.gmt")

metabo2 <- filter(top, set %in% names(kegg_metabosets))
metabo2_ord <- metabo2[order(metabo2$s.dist),]
metabo2_ord$s.dist <- abs(metabo2_ord$s.dist)

barplot(metabo2_ord$s.dist,horiz=TRUE,las=1,names.arg=metabo2_ord$set,
  main="Enriched Metabolic Pathways - KEGG",cex.names=0.6,xlab="Enrichment Score",
  col=rev(c(rep("red",nrow(filter(metabo2, s.dist > 0))),rep("blue",nrow(filter(metabo2, s.dist < 0))))),
  xlim=c(0,0.8))

```
Heatmaps for most enriched pathways in cell lines, on biopsy data.

```{r, complexheatmap13}

# see https://www.youtube.com/watch?v=ht1r34-ifVI for reference

dge3 <- dge2

# find corresponding human genes to devil ENS IDs
dge2$ENSSHAG <- rownames(dge2)
hm3 <- data.frame(ENSSHAG=hm2[,2],geneID=hm2[,3])
dge2 <- left_join(dge2, hm3, by="ENSSHAG")
dge2 <- unite(dge2,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ", remove=F)
dge2 <- unique(dge2)
rownames(dge2) <- dge2$ENSSHAG_geneID

# only plot significant DEGs
dge2sig <- filter(dge2, padj < 0.05)
dge2sig <- dge2sig[order(dge2sig$log2FoldChange, decreasing=TRUE),]

mx <- dge2sig[,7:ncol(dge2sig)] # get normalised counts
 
mx.scaled <- t(apply(mx[,1:4], 1, scale)) # center and scale each column (Z-score)
colnames(mx.scaled) <- colnames(mx[,1:4])
colnames(mx.scaled) <- c("DFT1 1", "DFT1 2", "DFT2 1", "DFT2 2")
log2FC <- as.matrix(dge2sig$log2FoldChange)
rownames(log2FC) <- rownames(dge2sig)
colnames(log2FC) <- "logFC"
mean <- as.matrix(dge2sig$baseMean)
rownames(mean) <- rownames(dge2sig)
colnames(mean) <- "AveExpr"

col_logFC <- col_logFC <- colorRamp2(c(-20,0,20), hcl_palette = "Blue-Red 2")
#col_AveExpr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white","red"))

ha <- HeatmapAnnotation(summary=anno_summary(gp=gpar(fill="#D5D7E2"),height=unit(2, "cm")))

# ----- Top DEGs in REACTOME metabolic pathways ------

metabo <- c() # get all names of genes involved in metabolism
for(i in 1:length(genesets)){metabo <- c(metabo,genesets[[i]])}
metabo <- unique(metabo) # get rid of duplicates

temp <- c() # get ENSSHAG and gene ID
for(i in 1:length(metabo)){temp <- rbind(temp,filter(hm3, geneID==metabo[i]))}
metabolism <- c(t(unite(temp,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% metabolism)
metFC <- subset(log2FC, rownames(log2FC) %in% metabolism)
metAE <- subset(mean, rownames(mean) %in% metabolism)

# top 20 up & down
met40 <- rbind(head(met,20),tail(met,20))
met40FC <- as.matrix(c(head(metFC,20),tail(metFC,20)))
rownames(met40FC) <- rownames(met40)
colnames(met40FC) <- "logFC"
met40AE <- as.matrix(c(head(metAE,20),tail(metAE,20)))
rownames(met40AE) <- rownames(met40)
colnames(met40AE) <- "AveExpr"

custom_heatmap(met40, met40FC, met40AE, title = "TOP 40 - REACTOME metabolism")

# ----- All DEGs in REACTOME metabolic pathways ------

dge2_metabo_reactome <- filter(dge2, geneID %in% metabo)

make_volcano(dge2_metabo_reactome, "Biopsies - Metabolism genes - REACTOME")

# ----- Individual TOP REACTOME metabolic pathways -----

pathways <- metabo1$set
heatmaps <- vector(mode = "list", length = length(pathways))
for(i in 1:length(pathways)){
  
  metabo <- c() # get all names of genes involved in metabolism
  for(j in 1:length(genesets)){metabo <- c(metabo,genesets[[pathways[i]]])}
  metabo <- unique(metabo) # get rid of duplicates

  temp <- c() # get ENSSHAG and gene ID
  for(k in 1:length(metabo)){temp <- rbind(temp,filter(hm3, geneID==metabo[k]))}
  metabolism <- c(t(unite(temp,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

  met <- subset(mx.scaled, rownames(mx.scaled) %in% metabolism)
  metFC <- subset(log2FC, rownames(log2FC) %in% metabolism)
  metAE <- subset(mean, rownames(mean) %in% metabolism)

  # react_pathw[[i]] <- custom_heatmap(met, metFC, metAE, title = paste(pathways[i], "- REACTOME"))
  heatmaps[[i]] <- custom_heatmap(met, metFC, metAE, title = paste(pathways[i]))
  
}
```

```{r, complexheatmap2, fig.dim = c(4, 2.75)}
print(heatmaps[[1]])
```

```{r, complexheatmap3, fig.dim = c(4, 2.75)}
print(heatmaps[[2]])
```

```{r, complexheatmap4, fig.dim = c(4, 3)}
print(heatmaps[[3]])
```

```{r, complexheatmap5, fig.dim = c(4, 3.25)}
print(heatmaps[[4]])
```

```{r, complexheatmap6, fig.dim = c(4, 3.25)}
print(heatmaps[[5]])
```

```{r, complexheatmap7, fig.dim = c(4, 2.75)}
print(heatmaps[[6]])
```

```{r, complexheatmap8, fig.dim = c(4, 5.5)}
print(heatmaps[[7]])
```

```{r, complexheatmap9, fig.dim = c(4, 3.5)}
print(heatmaps[[8]])
```

```{r, complexheatmap10, fig.dim = c(4, 3.25)}
print(heatmaps[[9]])
```

```{r, complexheatmap12, fig.dim = c(4, 4)}
# ----- Other interesting pathways REACTOME -----

mm <- filter(hm3, geneID %in% genesets[["Glycolysis"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

custom_heatmap(met, metFC, metAE, title = "Glycolysis")
```

```{r, complexheatmap14, fig.dim = c(4, 6)}
mm <- filter(hm3, geneID %in% genesets[["Formation of ATP by chemiosmotic coupling"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

# custom_heatmap(met, metFC, metAE, title = "Formation of ATP by chemiosmotic coupling")
```

```{r, complexheatmap15, fig.dim = c(4, 2.5)}
mm <- filter(hm3, geneID %in% genesets[["Citric acid cycle (TCA cycle)"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

custom_heatmap(met, metFC, metAE, title = "Citric acid cycle (TCA cycle)")
```

```{r, complexheatmap16, fig.dim = c(4, 3.5)}

mm <- filter(hm3, geneID %in% genesets[["Respiratory electron transport, ATP synthesis by chemiosmotic coupling, and heat production by uncoupling proteins."]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

custom_heatmap(met, metFC, metAE, title = "Respiratory electron transport, \n ATP synthesis by chemiosmotic coupling, \n and heat production by uncoupling proteins.")
```

```{r, complexheatmap17, fig.dim = c(4, 4)}
mm <- filter(hm3, geneID %in% genesets[["DNA Repair"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

#custom_heatmap(met, metFC, metAE, title = "DNA Repair")

```

```{r, complexheatmap18, fig.dim = c(4, 2.75)}
mm <- filter(hm3, geneID %in% genesets[["Cholesterol biosynthesis"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

custom_heatmap(met, metFC, metAE, title = "Cholesterol biosynthesis")
```

```{r, complexheatmap19, fig.dim = c(4, 3.5)}
mm <- filter(hm3, geneID %in% genesets[["Detoxification of Reactive Oxygen Species"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

custom_heatmap(met, metFC, metAE, title = "Detoxification of Reactive Oxygen Species")
```

```{r, complexheatmap20, fig.dim = c(4, 2.75)}
mm <- filter(hm3, geneID %in% genesets[["Pentose phosphate pathway"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

custom_heatmap(met, metFC, metAE, title = "Pentose phosphate pathway")
```

```{r, complexheatmap21, fig.dim = c(4, 5.5)}
mm <- filter(hm3, geneID %in% genesets[["Fatty acid metabolism"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

custom_heatmap(met, metFC, metAE, title = "Fatty acid metabolism")
```

```{r, complexheatmap22, fig.dim = c(4, 2.5)}

mm <- filter(hm3, geneID %in% genesets[["Glutamate and glutamine metabolism"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

custom_heatmap(met, metFC, metAE, title = "Glutamate and glutamine metabolism")
```

```{r, complexheatmap23, fig.dim = c(4, 5.5)}

# ----- Top DEGs in KEGG metabolic pathways ------

metabo <- c()
for(i in 1:length(genesets2)){metabo <- c(metabo,genesets2[[i]])}
metabo <- unique(metabo)

temp <- c()
for(i in 1:length(metabo)){temp <- rbind(temp,filter(hm3, geneID==metabo[i]))}
metabolism <- c(t(unite(temp,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% metabolism)
metFC <- subset(log2FC, rownames(log2FC) %in% metabolism)
metAE <- subset(mean, rownames(mean) %in% metabolism)

# top 20 up & down
met40 <- rbind(head(met,20),tail(met,20))
met40FC <- as.matrix(c(head(metFC,20),tail(metFC,20)))
rownames(met40FC) <- rownames(met40)
colnames(met40FC) <- "logFC"
met40AE <- as.matrix(c(head(metAE,20),tail(metAE,20)))
rownames(met40AE) <- rownames(met40)
colnames(met40AE) <- "AveExpr"

custom_heatmap(met40, met40FC, met40AE, title = "TOP 40 - KEGG metabolism")

# ----- All DEGs in KEGG metabolic pathways ------

dge2_metabo_kegg <- filter(dge2, geneID %in% metabo)

make_volcano(dge2_metabo_kegg, "Biopsies - Metabolism genes - KEGG")

# ----- Individual TOP KEGG metabolic pathways -----

pathways <- metabo2$set

for(i in 1:length(pathways)){
  
  metabo <- c() # get all names of genes involved in metabolism
  for(j in 1:length(genesets2)){metabo <- c(metabo,genesets2[[pathways[i]]])}
  metabo <- unique(metabo) # get rid of duplicates

  temp <- c() # get ENSSHAG and gene ID
  for(k in 1:length(metabo)){temp <- rbind(temp,filter(hm3, geneID==metabo[k]))}
  metabolism <- c(t(unite(temp,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

  met <- subset(mx.scaled, rownames(mx.scaled) %in% metabolism)
  metFC <- subset(log2FC, rownames(log2FC) %in% metabolism)
  metAE <- subset(mean, rownames(mean) %in% metabolism)

  # react_pathw[[i]] <- custom_heatmap(met, metFC, metAE, title = paste(pathways[i], "- REACTOME"))
  print(custom_heatmap(met, metFC, metAE, title = paste(pathways[i], "- KEGG")))
  
}

```
```{r}
seb_genes <- filter(dge2, geneID %in% c("LICAM","CXCL13","GDNF", "NGF", "NRTN", "ARTN", "GFRA1", "NTF3", "CCL2", "CX3CL1"))

sig <- subset(seb_genes,padj<0.05)

# heatmap
mx <- sig[,7:10]
mx <- head(mx,30)
colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(as.matrix(mx),trace="none",scale="row",
  col=colfunc(25),mar=c(5,14),
  cexRow=0.8,cexCol=0.7)
```

```{r, Venn data}

allgenes <- rownames(dge)
sigdegs <- rownames(sig2)
updegs <- sig2_up
dndegs <- sig2_dn

metreactgenes <- rownames(dge2_metabo_reactome)
metsigreactdegs <- rownames(filter(dge2_metabo_reactome, padj < 0.05))
metsigreactupdegs <- rownames(filter(filter(dge2_metabo_reactome, padj < 0.05),log2FoldChange > 0))
metsigreactdndegs <- rownames(filter(filter(dge2_metabo_reactome, padj < 0.05),log2FoldChange < 0))
metkegggenes <- rownames(dge2_metabo_kegg)
metsigkeggdegs <- rownames(filter(dge2_metabo_kegg, padj < 0.05))
metsigkeggupdegs <- rownames(filter(filter(dge2_metabo_kegg, padj < 0.05),log2FoldChange > 0))
metsigkeggdndegs <- rownames(filter(filter(dge2_metabo_kegg, padj < 0.05),log2FoldChange < 0))

venn_biopsies <- list(allgenes=allgenes, sigdegs=sigdegs, updegs=updegs, dndegs=dndegs, metreactgenes=metreactgenes,
                      metsigreactdegs=metsigreactdegs, metsigreactupdegs=metsigreactupdegs, metsigreactdndegs=metsigreactdndegs,
                      metkegggenes=metkegggenes, metsigkeggdegs=metsigkeggdegs, metsigkeggupdegs=metsigkeggupdegs,
                      metsigkeggdndegs=metsigkeggdndegs)
saveRDS(venn_biopsies, file = "venn_biopsies.rds")
saveRDS(dge3, "~/dftd_RNAseq_annelise/dge/dge_biopsies.rds")

# venn.diagram(
#   x = list(alldegs, updegs, downdegs),
#   category.names = c("All DEGs" , "Up", "Down"),
#   filename = 'updown_venn_diagramm.png',
#   output=TRUE
# )

```

#### Term-gene graphs

```{r, tgg}

degenes <- data.frame(FC=dge2sig$log2FoldChange, ID=dge2sig$geneID)

paths <- c("Glycolysis",
           "Formation of ATP by chemiosmotic coupling",
           "Citric acid cycle (TCA cycle)",
           "Respiratory electron transport",
           "DNA Double-Strand Break Repair",
           "Detoxification of Reactive Oxygen Species",
           "Pentose phosphate pathway",
           "Fatty acid metabolism",
           "Glutamate and glutamine metabolism")

path_list <- list()

for(i in 1:nrow(metabo1)){
  path_temp <- metabo1$set[i]
  genes_temp <- genesets[[path_temp]]
  list_temp <- list(intersect(genes_temp,degenes$ID))
  path_list <- append(path_list, list_temp)
}

names(path_list) <- metabo1$set

degenes <- na.omit(data.frame(ID=dge2sig$geneID,FC=dge2sig$log2FoldChange))

mydf_2 <- degenes %>% group_by(ID) %>% summarize(FC=mean(FC))
mydf_2 <- data.frame(mydf_2)
mydf_2 <- mydf_2[-1,]
mydf_3 <- mydf_2[,-1]

FC <- setNames(mydf_3, mydf_2$ID)

cnetplot(path_list, foldChange=FC) 



path_list <- list()

for(i in 1:length(paths)){
  path_temp <- paths[i]
  genes_temp <- genesets[[path_temp]]
  list_temp <- list(intersect(genes_temp,degenes$ID))
  path_list <- append(path_list, list_temp)
}

names(path_list) <- paths

degenes <- na.omit(data.frame(ID=dge2sig$geneID,FC=dge2sig$log2FoldChange))

mydf_2 <- degenes %>% group_by(ID) %>% summarize(FC=mean(FC))
mydf_2 <- data.frame(mydf_2)
mydf_2 <- mydf_2[-1,]
mydf_3 <- mydf_2[,-1]

FC <- setNames(mydf_3, mydf_2$ID)

# TODO: show TCA+ETC and FOA+ROS
cnetplot(path_list, foldChange=FC, showCategory = 3) 

```

```{r, visualise pathways}
data("sbgn.xmls")
data(pathways.info)
input.pathways <- data.frame(findPathways(c(names(path_list),metabo1$set)))
SBGNview.obj <- SBGNview(
          gene.data = FC, 
          gene.id.type = "SYMBOL",
          input.sbgn = input.pathways$pathway.id,
          output.file = "biopsies", 
          output.formats =  c("svg"),
          min.gene.value = -10,
          max.gene.value = 10)
print(SBGNview.obj)
```


## Adding custom gene sets

```{r, customgs, fig.dim = c(4, 2.75)}

cell_compet_genes <- read_excel("../ref/gsea/Cell_compet_genes.xlsx")
names(cell_compet_genes)[names(cell_compet_genes) == "Ensembl_ID"] <- "Gene.stable.ID"

cell_compet_genes <- inner_join(cell_compet_genes,hm,by="Gene.stable.ID") # making sure these genes exist in devils

genesets$"Cell Competition" <- unique(cell_compet_genes$Gene.name)
genesets$"Cell Competition - Mechanical Competition" <- unique(filter(cell_compet_genes, Type == "Mechanical_Competition")$Gene.name)

metastasis_genes <- read_excel("../ref/gsea/Metastasis_genes.xlsx")
names(metastasis_genes)[names(metastasis_genes) == "Ensembl_ID"] <- "Gene.stable.ID"

metastasis_genes <- inner_join(metastasis_genes,hm,by="Gene.stable.ID") # making sure these genes exist in devils

genesets$"Metastasis" <- unique(metastasis_genes$Gene.name)
genesets$"Metastasis - promotors" <- unique(filter(metastasis_genes, Type == "Metastasis_promotor")$Gene.name)
genesets$"Metastasis - suppressors" <- unique(filter(metastasis_genes, Type == "Metastasis_supressor")$Gene.name)

# gsea
res_custom <- mitch_calc(m2, genesets, priority="effect",cores=16)

if ( !file.exists("mitch_custom_patchett.html") ) {
  mitch_report(res_custom, "mitch_custom_patchett.html")
}


cc <- filter(hm3, geneID %in% genesets[["Cell Competition"]])
cc <- c(t(unite(cc,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% cc)
metFC <- subset(log2FC, rownames(log2FC) %in% cc)
metAE <- subset(mean, rownames(mean) %in% cc)

custom_heatmap(met, metFC, metAE, title = "Cell competition")
```

```{r, customgs2, fig.dim = c(4, 4.5)}
meta <- filter(hm3, geneID %in% genesets[["Metastasis"]])
meta <- c(t(unite(meta,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

met <- subset(mx.scaled, rownames(mx.scaled) %in% meta)
metFC <- subset(log2FC, rownames(log2FC) %in% meta)
metAE <- subset(mean, rownames(mean) %in% meta)

custom_heatmap(met, metFC, metAE, title = "Metastasis")

```