---
title: "Re-analysing Patchett 2020 data"
author: "Anne-Lise GÃ©rard"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 5
    fig_height: 5
theme: cosmo
---

Source codes: https://github.com/Anne-LiseGerard/dftd_rnaseq
Data: https://doi.org/10.1007/s00018-019-03259-2

```{r,pkg}

suppressPackageStartupMessages({
    library("reshape2")
    library("gplots")
    library("DESeq2")
    library("mitch")
    library("limma")
    library("kableExtra")
    library("dplyr")
    library("tidyr")
    library("ComplexHeatmap")
    library("RColorBrewer")
    library("circlize")
    library("pathview")
    library("stringr")
    library("readxl")
    library("grid")
    library("ggplot2")
    library("viridis")
    library("enrichplot")
})

knitr::opts_chunk$set(dev = 'svg')

```

## Background

Goal: re-analyse previously published DFTD tumour biopsy transcriptomes and compare to DFTD cell line transcriptomes, focusing on genes involved in metabolism. Below is a list of the samples for this project.

```{r, ss}

ss <- read.table("../ss_patchett.txt",sep="\t",fill=TRUE,header=TRUE)
ss$DFT <- as.factor(ss$DFT)

ss %>%
  kbl(caption="Sample sheet for all samples") %>%
  kable_paper("hover", full_width = F)

```

## Functions

```{r, func}
# volcanoplots
make_volcano <- function(de,name) {
    sig <- subset(de,padj<0.05)
    N_SIG=nrow(sig)
    N_UP=nrow(subset(sig,log2FoldChange>0))
    N_DN=nrow(subset(sig,log2FoldChange<0))
    DET=nrow(de)
    HEADER=paste(N_SIG,"@5%FDR,", N_UP, "up", N_DN, "dn", DET, "detected")
    plot(de$log2FoldChange,-log10(de$padj),cex=1,pch=19,col="#D5D7E2",
        main=name, xlab="log2 FC", ylab="-log10 pval")
    mtext(HEADER)
    grid()
    points(sig$log2FoldChange,-log10(sig$padj),cex=1,pch=19,col="#5297A7")
}

#heatmaps
custom_heatmap <- function(zscore, fc, aveexpr, title) {
  h1 <- Heatmap(zscore, cluster_rows = F,
              column_labels = colnames(zscore), name="Z-score",
              cluster_columns = T, 
              column_names_gp = gpar(fontsize = 7, fontface="bold"),
              col=colorRamp2(c(-1.5,0,1.5), hcl_palette = "Blue-Red 2"),
              heatmap_legend_param = list(title = "Z-score", at = seq(-1.5, 1.5, 0.5)))
  h2 <- Heatmap(fc, row_labels = rownames(fc), row_names_gp = gpar(fontsize = 6),
              cluster_rows = F, name="logFC", col = col_logFC,column_names_gp = gpar(fontsize =7,fontface="bold"),
              cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
                grid.text(round(as.numeric(fc[i, j],2),2), x, y,gp=gpar(fontsize=6))
              }, column_title = title, 
              heatmap_legend_param = list(title = "logFC", at = seq(-20, 20, 10)))

  h <- h1 + h2
  h
}

```

## Load data

Import data from the aligner. 

```{r, import}

tmp <- read.table("../fastq/3col_patchett.tsv.gz")
tmp$V1 <- gsub('.fastq-trimmed.fastq','',tmp$V1)
x <- as.data.frame(acast(tmp, V2~V1, value.var="V3",fun.aggregate=sum))

tmp_old <- read.table("../fastq/3col_old.tsv.gz")
tmp_old$V1 <- gsub('.fastq-trimmed.fastq','',tmp_old$V1)
x_old <- as.data.frame(acast(tmp_old, V2~V1, value.var="V3",fun.aggregate=sum))

```

Load gene names.

```{r, genenames}

gn <- read.table("../ref/Sarcophilus_harrisii.mSarHar1.11.cdna+ncrna.genenames.tsv",fill=TRUE)
gn <- gn[order(gn$V1),]

gn_old <- read.table("../ref/Sarcophilus_harrisii.DEVIL7.0.cdna+ncrna.genenames.tsv",fill=TRUE)
gn_old <- gn_old[order(gn_old$V1),]

```

Load homology table.

```{r, homology}

hm <- read.table("../ref/mart_export_ensembl109_2023-07-14.txt",sep="\t",header=TRUE)

hm_old <- read.table("../ref/mart_export_ensembl101_2023-09-06.txt",sep="\t",header=TRUE)

```

Now need to collapse transcript data to genes.

```{r, collapse}

x$gene <- paste(gn$V2,gn$V3)

y <- aggregate(. ~ gene,x,sum)

rownames(y) <- y$gene
y$gene = NULL


x_old$gene <- paste(gn_old$V2,gn_old$V3)

y_old <- aggregate(. ~ gene,x_old,sum)

rownames(y_old) <- y_old$gene
y_old$gene = NULL

```

## Quality control

Samples with <1M reads should be omitted. Will also round values to integers. Using the new genome yields a higher number of mapped reads (this data was originally mapped on the DEVIL 7.0 assembly in Patchett 2019).

```{r, qc1}

cs <- colSums(y)
cs <- cs[order(cs)]

par(mar=c(5,10,5,2))
barplot(cs,main="All samples",horiz=TRUE,las=1)
abline(v=1e7,col="red",lty=2)

y <- round(y)


cs_old <- colSums(y_old)
cs_old <- cs_old[order(cs_old)]

par(mar=c(5,10,5,2))
barplot(cs_old,main="All samples - old genome ",horiz=TRUE,las=1, xli=c(0,4e+07))
abline(v=1e7,col="red",lty=2)

y_old <- round(y_old)

```

## MDS

This will help us to visualise the sources of variability in the overall dataset. Plot MDS. Also fix the sample names. Plot MDS by DFT. Plot MDS by cell lines and biiopsies. Decided to only include biopsies for rest of analysis.

```{r, plotmds1}

par(mar = c(5.1, 4.1, 4.1, 2.1))

colnames(y) <- ss$sample_id

saveRDS(y, file = "y_biopsies.rds")

cs <- colSums(y)
cs <- cs[order(cs)]

par(mar=c(5,10,5,2))
barplot(cs,main="All samples",horiz=TRUE,las=1)


par(mar = c(5.1, 4.1, 4.1, 2.1))

colnames(y_old) <- ss$sample_id[1:8]

cs_old <- colSums(y_old)
cs_old <- cs_old[order(cs_old)]

par(mar=c(5,10,5,2))
barplot(cs_old,main="All samples - old genome",horiz=TRUE,las=1, xlim=c(0,4e+07))


par(mar = c(5.1, 4.1, 4.1, 2.1) )

cols <- ss$DFT
cols <- gsub("DFT1","#B3B9DF",cols)
cols <- gsub("DFT2","#DDAFB7",cols)
mymds <- plotMDS(y,plot=FALSE)

mymds_old <- plotMDS(y_old,plot=FALSE)

# fix the xlims
XMIN=min(mymds$x)
XMAX=max(mymds$x)
XMID=(XMAX+XMIN)/2
XMIN <- XMID + (XMIN-XMID)*1.1
XMAX <- XMID+(XMAX-XMID)*1.1
plotMDS(mymds,pch=17,cex=3,col=cols,main="MDS plot",xlim=c(XMIN,XMAX))
text(mymds,labels=colnames(y))
mtext("blue=DFT1, pink=DFT2")

plotMDS(mymds_old,pch=17,cex=3,col=cols,main="MDS plot - old genome",xlim=c(XMIN,XMAX))
text(mymds_old,labels=colnames(y_old))
mtext("blue=DFT1, pink=DFT2")

y_DFT1 <- y[,colnames(y) %in% c("sha_DFT1_1-1","sha_DFT1_1-2","sha_DFT1_2-1","sha_DFT1_2-2",
                                "C5065_UT_01", "C5065_UT_02")]
y_DFT2 <- y[,colnames(y) %in% c("sha_DFT2_1-1","sha_DFT2_1-2","sha_DFT2_2-1","sha_DFT2_2-2",
                                "sha_DFT2_RV_1-1","sha_DFT2_RV_1-2","sha_DFT2_RV_2-1","sha_DFT2_RV_2-2")] 

y_biopsy <- y[,colnames(y) %in% c("sha_DFT1_1-1","sha_DFT1_1-2","sha_DFT1_2-1","sha_DFT1_2-2",
                                  "sha_DFT2_1-1","sha_DFT2_1-2","sha_DFT2_2-1","sha_DFT2_2-2")]
y_cline <- y[,colnames(y) %in% c( "C5065_UT_01", "C5065_UT_02",
                                "sha_DFT2_RV_1-1","sha_DFT2_RV_1-2","sha_DFT2_RV_2-1","sha_DFT2_RV_2-2")] 

mdsDFT1 <- plotMDS(y_DFT1,plot=FALSE)
plotMDS(mdsDFT1,pch=17,cex=3,col="#B3B9DF",main="MDS plot",xlim=c(XMIN,XMAX))
text(mdsDFT1,labels=colnames(y_DFT1))
mtext("DFT1")

mdsDFT2 <- plotMDS(y_DFT2,plot=FALSE)
plotMDS(mdsDFT2,pch=17,cex=3,col="#DDAFB7",main="MDS plot",xlim=c(XMIN,XMAX))
text(mdsDFT2, labels=colnames(y_DFT2))
mtext("DFT2")

mds_biopsy <- plotMDS(y_biopsy,plot=FALSE)
plotMDS(mds_biopsy,pch=17,cex=3,col=cols,main="Tumour biopsies",xlim=c(XMIN,XMAX))
text(mds_biopsy,labels=colnames(y_biopsy))
mtext("Tumour biopsies: blue=DFT1, pink=DFT2")

mds_cline <- plotMDS(y_cline,plot=FALSE)
plotMDS(mds_cline,pch=17,cex=3,col=cols,main="MDS plot",xlim=c(XMIN,XMAX))
text(mds_cline,labels=colnames(y_cline))
mtext("Cell lines: blue=DFT1, pink=DFT2")

```

## DESeq2

Sum replicates then run differential expression analysis. For now, ignore alignment on old genome and cell line data.

```{r, de}

y_cell <- y[,9:14]
y <- y[,1:8]

# combine replicates for each sample
DFT1_1 <- rowSums(y[,ss$sample=="DFT1_1"])
DFT1_2 <- rowSums(y[,ss$sample=="DFT1_2"])
DFT2_1 <- rowSums(y[,ss$sample=="DFT2_1"])
DFT2_2 <- rowSums(y[,ss$sample=="DFT2_2"])

y2 <- data.frame(DFT1_1,DFT1_2,DFT2_1,DFT2_2)
ss2 <- as.data.frame(colnames(y2))
colnames(ss2) <- "sample"
ss2$DFT <- factor(c("DFT1","DFT1","DFT2","DFT2"))

y2 <- y2[which(rowMeans(y2)>10),] # remove genes with average count < 10
dim(y2)

dds <- DESeqDataSetFromMatrix(countData = y2 , colData = ss2, design = ~ DFT )
dds2 <- dds
res <- DESeq(dds)
z <- results(res)
vsd <- vst(dds)
zz<-cbind(as.data.frame(z),assay(vsd))
dge<-as.data.frame(zz[order(zz$pvalue),])
dge2 <- dge
sig <- subset(dge,padj<0.05)
sig2 <- sig
sig2_up <- rownames(subset(sig,log2FoldChange>0))
sig2_dn <- rownames(subset(sig,log2FoldChange<0))
length(sig2_up)
length(sig2_dn)

```

```{r,deviz}


make_volcano(dge2,"Tumour biopsies")

sig[1:50,1:6] %>%
  kbl(caption="Tumour biopsies") %>%
  kable_paper("hover", full_width = F)

write.table(dge2,file="dge3.tsv",sep="\t")


```

```{r,de2vizheatmap,fig.dim = c(4, 6)}

# top 20 up & top 20 down
sig <- sig[order(sig$log2FoldChange, decreasing=TRUE),]

sig_noNA <- sig
sig_noNA$genes <- rownames(sig_noNA)
sig_noNA <- filter(sig_noNA, !grepl(' NA', genes))

top <- rbind(head(sig_noNA,20),tail(sig_noNA,20))

mx <- top[,7:ncol(top)] # get normalised counts

mx.scaled <- t(apply(mx[,1:4], 1, scale)) # center and scale each column (Z-score)
colnames(mx.scaled) <- colnames(mx[,1:4])
colnames(mx.scaled) <- c("DFT1 1", "DFT1 2", "DFT2 1", "DFT2 2")
log2FC <- as.matrix(top$log2FoldChange)
rownames(log2FC) <- rownames(top)
colnames(log2FC) <- "logFC"
mean <- as.matrix(top$baseMean)
rownames(mean) <- rownames(top)
colnames(mean) <- "AveExpr"


col_logFC <- colorRamp2(c(-20,0,20), hcl_palette = "Blue-Red 2")
custom_heatmap <- function(zscore, fc, aveexpr, title) {
  h1 <- Heatmap(zscore, cluster_rows = F,
              column_labels = colnames(zscore), name="Z-score",
              cluster_columns = T, 
              column_names_gp = gpar(fontsize = 7, fontface="bold"),
              col=colorRamp2(c(-1.5,0,1.5), hcl_palette = "Blue-Red 2"),
              heatmap_legend_param = list(title = "Z-score", at = seq(-1.5, 1.5, 0.5)))
  h2 <- Heatmap(fc, row_labels = rownames(fc), row_names_gp = gpar(fontsize = 6),
              cluster_rows = F, name="logFC", col = col_logFC,column_names_gp = gpar(fontsize =7,fontface="bold"),
              cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
                grid.text(round(as.numeric(fc[i, j],2),2), x, y,gp=gpar(fontsize=6))
              }, column_title = title, 
              heatmap_legend_param = list(title = "logFC", at = seq(-20, 20, 10)))

  h <- h1 + h2 
  h
}

ha <- HeatmapAnnotation(summary=anno_summary(gp=gpar(fill="#D5D7E2"),height=unit(2, "cm")))

rownames(mx.scaled) <- gsub("^.{0,21}", "", rownames(mx.scaled))
rownames(log2FC) <- gsub("^.{0,21}", "", rownames(log2FC))


top40 <- custom_heatmap(mx.scaled, log2FC, mean, title = "Tumour biopsies")

top40

saveRDS(top40, "top40_biopsies.RDS")

```
## Enrichment analysis

Need to get human homologs of these genes. These were obtained from Ensembl v109 biomart.

```{r}

rownames(dge2) <- sapply(strsplit(rownames(dge2),"\\."),"[[",1)

hm2 <- hm[hm$Tasmanian.devil.gene.stable.ID != "",]

gt <- hm2[,2:3]
length(unique(gt$Tasmanian.devil.gene.stable.ID))
length(unique(gt$Gene.name))

for(i in 1:length(gt$Gene.name)){
  if(gt$Gene.name[i]==""){gt$Gene.name[i] <- gt$Tasmanian.devil.gene.stable.ID[i]}
}

saveRDS(gt, "~/dftd_RNAseq_annelise/dge/gt.rds")

```

Now run mitch for DGE2. The pathways are sourced from Reactome 7th July 2023. Plot significant pathways, top 20 and metabolism related pathways.

```{r,mitch1, fig.dim = c(12, 22)}

genesets <- gmt_import("../ref/ReactomePathways_2023-07-14.gmt")

m2 <- mitch_import(dge2, DEtype="deseq2", geneTable=gt)
head(m2)

res2 <- mitch_calc(m2, genesets, priority="effect",cores=16)

if ( !file.exists("mitch_biopsies_reactome.html") ) {
  mitch_report(res2, "mitch_biopsies_reactome.html")
}

top <- subset(res2$enrichment_result,p.adjustANOVA<0.05)
saveRDS(top, "~/dftd_RNAseq_annelise/dge/gsea_biopsies.rds")

top %>%
  kbl(caption="Enriched pathways") %>%
  kable_paper("hover", full_width = F)

ggplot(top, aes(x = s.dist, y = reorder(set, s.dist), color = p.adjustANOVA, size = setSize)) + 
  geom_point(stat = 'identity') + 
  scale_color_viridis(option="mako", direction=1,limits = c(0, 0.05)) +
  xlab("s dist") + ylab("pathway") + ggtitle("Reactome ALL") + 
  theme_bw()

```
```{r,mitch2, fig.dim = c(8, 6)}
top20 <- rbind(slice_max(top, order_by=s.dist, n=10), slice_min(top, order_by=s.dist, n=10))

ggplot(top20, aes(x = s.dist, y = reorder(set, s.dist), color = p.adjustANOVA, size = setSize)) + 
  geom_point(stat = 'identity') + 
  scale_color_viridis(option="mako", direction=1,limits = c(0, 0.05)) +
  xlab("s dist") + ylab("pathway") + ggtitle("Reactome top 20") + 
  theme_bw()
```
```{r,mitch3, fig.dim = c(12, 6)}
react_metabosets <- readRDS("../ref/reactome_metabo.rds")

metabo1 <- filter(top, set %in% names(react_metabosets))
saveRDS(metabo1, "~/dftd_RNAseq_annelise/dge/gsea_biopsies_metabo.rds")

ggplot(metabo1 , aes(x = s.dist, y = reorder(set, s.dist), color = p.adjustANOVA, size = setSize)) + 
  geom_point(stat = 'identity') + 
  scale_color_viridis(option="mako", direction=1,limits = c(0, 0.05)) +
  xlab("s dist") + ylab("pathway") + ggtitle("Reactome metabolism") + 
  theme_bw()

```

## Visualisation

Heatmaps. First plot enriched pathways, then plot pathways of interest for which experimental data is available.


```{r, complexheatmap13, fig.dim = c(4, 6)}

# see https://www.youtube.com/watch?v=ht1r34-ifVI for reference

hm3 <- data.frame(ENSSHAG=hm2[,2],geneID=hm2[,3])
dge2 <- dge
dge2$ensembl <- rownames(dge2)
split <- strsplit(dge2$ensembl, split = " ")
ensembl <- c()
gene_name <- c()
for (i in 1:length(split)){
  ensembl <- append(ensembl, split[[i]][1]) 
  gene_name <- append(gene_name, split[[i]][2]) 
}
dge2$ensembl <- ensembl
dge2$gene_name <- gene_name

dge3 <- dge2

# only plot significant DEGs
dge3sig <- filter(dge3, padj < 0.05)
dge3sig <- dge3sig[order(dge3sig$log2FoldChange, decreasing=TRUE),]

mx <- dge3sig[,7:ncol(dge3sig)] # get normalised counts
 
mx.scaled <- t(apply(mx[,1:4], 1, scale)) # center and scale each column (Z-score)
colnames(mx.scaled) <- c("DFT1 1", "DFT1 2", "DFT2 1", "DFT2 2")
log2FC <- as.matrix(dge3sig$log2FoldChange)
rownames(log2FC) <- rownames(dge3sig)
colnames(log2FC) <- "logFC"
mean <- as.matrix(dge3sig$baseMean)
rownames(mean) <- rownames(dge3sig)
colnames(mean) <- "AveExpr"

col_logFC <- colorRamp2(c(-20,0,20), hcl_palette = "Blue-Red 2")
#col_AveExpr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white","red"))

ha <- HeatmapAnnotation(summary=anno_summary(gp=gpar(fill="#D5D7E2"),height=unit(1.5, "cm")))

# ----- Top DEGs in REACTOME metabolic pathways ------

metabo <- c() # get all names of genes involved in metabolism
for(i in 1:length(genesets)){metabo <- c(metabo,genesets[[i]])}
metabo <- unique(metabo) # get rid of duplicates

temp <- c() # get ENSSHAG and gene ID
for(i in 1:length(metabo)){temp <- rbind(temp,filter(hm3, geneID==metabo[i]))}
metabolism <- c(t(unite(temp,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

# collapse transcript IDs to gene IDs

rownames(mx.scaled) <- sub('\\..*\\s', ' ', rownames(mx.scaled))
rownames(log2FC) <- sub('\\..*\\s', ' ', rownames(log2FC))
rownames(mean) <- sub('\\..*\\s', ' ', rownames(mean))
dge3$ensembl <- sub('\\..*', '', dge3$ensembl)
rownames(dge3) <- sub('\\..*\\s', ' ', rownames(dge3))

rownames(mx.scaled) <- gsub("^.{0,19}", "", rownames(mx.scaled))
rownames(log2FC) <- gsub("^.{0,19}", "", rownames(log2FC))

metabolism <- gsub("^.{0,19}", "", metabolism)

met <- subset(mx.scaled, rownames(mx.scaled) %in% metabolism)
metFC <- subset(log2FC, rownames(log2FC) %in% metabolism)
metAE <- subset(mean, rownames(mean) %in% metabolism)

# top 20 up & down
met40 <- rbind(head(met,20),tail(met,20))
met40FC <- as.matrix(c(head(metFC,20),tail(metFC,20)))
rownames(met40FC) <- rownames(met40)
colnames(met40FC) <- "logFC"
met40AE <- as.matrix(c(head(metAE,20),tail(metAE,20)))
#rownames(met40AE) <- rownames(met40)
#colnames(met40AE) <- "AveExpr"

custom_heatmap(met40, met40FC, met40AE, title = "TOP 40 - REACTOME metabolism")

# ----- All DEGs in REACTOME metabolic pathways ------

dge3_metabo_reactome <- filter(dge3, gene_name %in% metabo)

make_volcano(dge3_metabo_reactome, "Cell lines - Metabolism genes - REACTOME")

# ----- Individual TOP REACTOME metabolic pathways -----

pathways <- metabo1$set
heatmaps <- vector(mode = "list", length = length(pathways))
for(i in 1:length(pathways)){
  
  metabo <- c() # get all names of genes involved in metabolism
  for(j in 1:length(genesets)){metabo <- c(metabo,genesets[[pathways[i]]])}
  metabo <- unique(metabo) # get rid of duplicates

  temp <- c() # get ENSSHAG and gene ID
  for(k in 1:length(metabo)){temp <- rbind(temp,filter(hm3, geneID==metabo[k]))}
  metabolism <- c(t(unite(temp,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))
  
  metabolism <- gsub("^.{0,19}", "", metabolism)
  
  met <- subset(mx.scaled, rownames(mx.scaled) %in% metabolism)
  metFC <- subset(log2FC, rownames(log2FC) %in% metabolism)
  metAE <- subset(mean, rownames(mean) %in% metabolism)

  # react_pathw[[i]] <- custom_heatmap(met, metFC, metAE, title = paste(pathways[i], "- REACTOME"))
  heatmaps[[i]] <- custom_heatmap(met, metFC, metAE, title = paste(pathways[i]))
  
}

```

```{r, complexheatmap2, fig.dim = c(4, 2.75)}
print(heatmaps[[1]])
```

```{r, complexheatmap3, fig.dim = c(4, 2.75)}
print(heatmaps[[2]])

saveRDS(heatmaps[[2]], "cyst_b.rds")
```

```{r, complexheatmap4, fig.dim = c(4, 3)}
print(heatmaps[[3]])
```

```{r, complexheatmap5, fig.dim = c(4, 3.25)}
print(heatmaps[[4]])
```

```{r, complexheatmap6, fig.dim = c(4, 3.25)}
print(heatmaps[[5]])
```

```{r, complexheatmap7, fig.dim = c(4, 2.75)}
print(heatmaps[[6]])
```

```{r, complexheatmap8, fig.dim = c(4, 5.5)}
print(heatmaps[[7]])

saveRDS(heatmaps[[7]], "nucl_b.rds")
```

```{r, complexheatmap9, fig.dim = c(4, 3.5)}
print(heatmaps[[8]])
```

```{r, complexheatmap10, fig.dim = c(4, 3.25)}
print(heatmaps[[9]])
```

```{r, complexheatmap12, fig.dim = c(4, 4)}
# ----- Other interesting pathways REACTOME -----

mm <- filter(hm3, geneID %in% genesets[["Glycolysis"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

glyco_p <- custom_heatmap(met, metFC, title = "Tumour biopsies")

glyco_p

saveRDS(glyco_p, "glyco_p2.rds")
```

```{r, complexheatmap121, fig.dim = c(4, 4)}
# ----- Other interesting pathways REACTOME -----

mm <- filter(hm3, geneID %in% genesets[["Cholesterol biosynthesis"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

glyco_p <- custom_heatmap(met, metFC, title = "Cholesterol biosynthesis")

glyco_p

saveRDS(glyco_p, "chol_b.rds")
```

```{r, complexheatmap1211, fig.dim = c(4, 4)}
# ----- Other interesting pathways REACTOME -----

mm <- filter(hm3, geneID %in% genesets[["Sulfur amino acid metabolism"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

glyco_p <- custom_heatmap(met, metFC, title = "Sulfur amino acid metabolism")

glyco_p

saveRDS(glyco_p, "sulf_b.rds")
```

```{r, complexheatmap14, fig.dim = c(4, 6)}
mm <- filter(hm3, geneID %in% genesets[["Formation of ATP by chemiosmotic coupling"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

#mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, title = "Formation of ATP by chemiosmotic coupling")
```

```{r, complexheatmap15, fig.dim = c(4, 2.5)}
mm <- filter(hm3, geneID %in% genesets[["Citric acid cycle (TCA cycle)"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, title = "Citric acid cycle (TCA cycle)")
```

```{r, complexheatmap16, fig.dim = c(4, 3.5)}

mm <- filter(hm3, geneID %in% genesets[["Respiratory electron transport, ATP synthesis by chemiosmotic coupling, and heat production by uncoupling proteins."]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, title = "Respiratory electron transport, \n ATP synthesis by chemiosmotic coupling, \n and heat production by uncoupling proteins.")
```

```{r, complexheatmap17, fig.dim = c(4, 4)}
mm <- filter(hm3, geneID %in% genesets[["DNA Repair"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, title = "DNA Repair")

```

```{r, complexheatmap18, fig.dim = c(4, 2.75)}
mm <- filter(hm3, geneID %in% genesets[["Cholesterol biosynthesis"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, title = "Cholesterol biosynthesis")
```

```{r, complexheatmap19, fig.dim = c(4, 3.5)}
mm <- filter(hm3, geneID %in% genesets[["Detoxification of Reactive Oxygen Species"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

ROS_p2 <- custom_heatmap(met, metFC, metAE, title = "Tumour biopsies")

ROS_p2 

saveRDS(ROS_p2, "ROS_p2.rds")
```

```{r, complexheatmap20, fig.dim = c(4, 2.75)}
mm <- filter(hm3, geneID %in% genesets[["Pentose phosphate pathway"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, title = "Pentose phosphate pathway")
```

```{r, complexheatmap21, fig.dim = c(4, 5.5)}
mm <- filter(hm3, geneID %in% genesets[["Fatty acid metabolism"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, title = "Fatty acid metabolism")
```

```{r, complexheatmap22, fig.dim = c(4, 2.5)}

mm <- filter(hm3, geneID %in% genesets[["Glutamate and glutamine metabolism"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, title = "Glutamate and glutamine metabolism")
```

## Adding custom gene sets

Two custom gene sets were added to the previous analysis containing cell competition & metastasis related genes based on a literature search. 

```{r, customgs, fig.dim = c(4, 2.75)}

cell_compet_genes <- read_excel("../ref/gsea/Cell_compet_genes.xlsx")
names(cell_compet_genes)[names(cell_compet_genes) == "Ensembl_ID"] <- "Gene.stable.ID"

cell_compet_genes <- inner_join(cell_compet_genes,hm,by="Gene.stable.ID") # making sure these genes exist in devils

genesets$"Cell Competition" <- unique(cell_compet_genes$Gene.name)
genesets$"Cell Competition - Mechanical Competition" <- unique(filter(cell_compet_genes, Type == "Mechanical_Competition")$Gene.name)

metastasis_genes <- read_excel("../ref/gsea/Metastasis_genes.xlsx")
names(metastasis_genes)[names(metastasis_genes) == "Ensembl_ID"] <- "Gene.stable.ID"

metastasis_genes <- inner_join(metastasis_genes,hm,by="Gene.stable.ID") # making sure these genes exist in devils

genesets$"Metastasis" <- unique(metastasis_genes$Gene.name)
genesets$"Metastasis - promotors" <- unique(filter(metastasis_genes, Type == "Metastasis_promotor")$Gene.name)
genesets$"Metastasis - suppressors" <- unique(filter(metastasis_genes, Type == "Metastasis_supressor")$Gene.name)

res_custom <- mitch_calc(m2, genesets, priority="effect",cores=16)

if ( !file.exists("mitch_custom_patchett.html") ) {
  mitch_report(res_custom, "mitch_custom_patchett.html")
}

cc <- filter(hm3, geneID %in% genesets[["Cell Competition"]])
cc <- c(t(unite(cc,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

cc <- gsub("^.{0,19}", "", cc)

met <- subset(mx.scaled, rownames(mx.scaled) %in% cc)
metFC <- subset(log2FC, rownames(log2FC) %in% cc)

custom_heatmap(met, metFC, title = "Cell competition")
```

```{r, customgs2, fig.dim = c(4, 4.5)}
meta <- filter(hm3, geneID %in% genesets[["Metastasis"]])
meta <- c(t(unite(meta,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

meta <- gsub("^.{0,19}", "", meta)

met <- subset(mx.scaled, rownames(mx.scaled) %in% meta)
metFC <- subset(log2FC, rownames(log2FC) %in% meta)

custom_heatmap(met, metFC, title = "Metastasis")

```

## Session information

For reproducibility.

```{r,session}

sessionInfo()

```