---
title: "Examining global gene expression in DFT1 and DFT2 cells"
author: "Anne-Lise GÃ©rard"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: yes
    code_folding: hide
    fig_width: 5
    fig_height: 5
theme: cosmo
---

Source codes: https://github.com/Anne-LiseGerard/dftd_rnaseq

```{r, pkg}

suppressPackageStartupMessages({
    library("reshape2")
    library("gplots")
    library("DESeq2")
    library("mitch")
    library("limma")
    library("kableExtra")
    library("dplyr")
    library("tidyr")
    library("ComplexHeatmap")
    library("RColorBrewer")
    library("circlize")
    library("pathview")
    library("stringr")
    library("grid")
    library("VennDiagram")
    library("enrichplot")
    library("ggplot2")
    library("viridis")
    library("readxl")
    library("cowplot")
    library("grid")
})

knitr::opts_chunk$set(dev = 'svg')

```

## Background

The goal of our study is to compare the expression of genes related to metabolism of DFT1 cell lines to DFT2 cell lines. For this, we have generated transcriptomes for 3 DFT1 and 3 DFT2 cell lines, each in triplicate.

```{r,ss1}

ss <- read.table("../ss.tsv",sep="\t",fill=TRUE,header=TRUE)
ss$DFT <- as.factor(ss$DFT)
ss$clone <- sapply(strsplit(ss$ClientID,"_"),"[[",1)

ss %>%
  kbl(caption="Sample sheet for all samples") %>%
  kable_paper("hover", full_width = F)

```

We are interested in comparing all DFT1 samples against all DFT2 samples. We will use mSarHar 1.11 from Ensembl v109 for the reference transcriptome.

## Functions

```{r,func}
# volcanoplots
make_volcano <- function(de,name) {
    sig <- subset(de,padj<0.05)
    N_SIG=nrow(sig)
    N_UP=nrow(subset(sig,log2FoldChange>0))
    N_DN=nrow(subset(sig,log2FoldChange<0))
    DET=nrow(de)
    HEADER=paste(N_SIG,"@5%FDR,", N_UP, "up", N_DN, "dn", DET, "detected")
    plot(de$log2FoldChange,-log10(de$padj),cex=1,pch=19,col="#D5D7E2",
        main=name, xlab="log2 FC", ylab="-log10 pval")
    mtext(HEADER)
    grid()
    points(sig$log2FoldChange,-log10(sig$padj),cex=1,pch=19,col="#5297A7")
}

# heatmaps
custom_heatmap <- function(zscore, fc, aveexpr, title) {
  h1 <- Heatmap(zscore, cluster_rows = F,
              column_labels = colnames(zscore), name="Z-score",
              cluster_columns = T, 
              column_names_gp = gpar(fontsize = 7, fontface="bold"),
              col=colorRamp2(c(-1.5,0,1.5), hcl_palette = "Blue-Red 2"),
              heatmap_legend_param = list(title = "Z-score", at = seq(-1.5, 1.5,0.5)))
  h2 <- Heatmap(fc, row_labels = rownames(fc), row_names_gp = gpar(fontsize = 6),
              cluster_rows = F, name="logFC", col = col_logFC,column_names_gp = gpar(fontsize =7,fontface="bold"),
              cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
                grid.text(round(as.numeric(fc[i, j],2),2), x, y,gp=gpar(fontsize=6))
              }, column_title = title, 
              heatmap_legend_param = list(title = "logFC", at = seq(-20, 20, 10)))

  h <- h1 + h2 
  h
}


```

## Load data

Here we load the data in from the aligner.

```{r,import}

tmp <- read.table("../fastq/3col.tsv.gz")
x <- as.data.frame(acast(tmp, V2~V1, value.var="V3",fun.aggregate=sum))
dim(x)

```

Load gene names.

```{r, genenames}

gn <- read.table("../ref/Sarcophilus_harrisii.mSarHar1.11.cdna+ncrna.genenames.tsv",fill=TRUE)

gn <- gn[order(gn$V1),]

dim(gn)

```

Load homology map.

```{r,homology}

hm <- read.table("../ref/mart_export_ensembl109_2023-07-14.txt",sep="\t",header=TRUE)

```

Now need to collapse transcript data to genes.

```{r,collapse}

x$gene <- paste(gn$V2,gn$V3)

y <- aggregate(. ~ gene,x,sum)

rownames(y) <- y$gene
y$gene = NULL

dim(y)

```

## Quality control

Samples with <1M reads should be omitted. Will also round values to integers.

```{r,qc1}

cs <- colSums(y)
cs <- cs[order(cs)]

par(mar=c(5,10,5,2))
barplot(cs,main="All samples",horiz=TRUE,las=1)
abline(v=1e7,col="red",lty=2)

y <- round(y)

```

## MDS

This will help us to visualise the sources of variability in the overall dataset. Fix sample names. Plot MDS of all samples then plot MDS by DFT.

```{r, plotmds1}

y <- y[,colnames(y) != "DEA5-4NEG"]

ss <- ss[ss$ClientID != "DEA4_6NEG",]

colnames(y) <- sapply(strsplit(ss$ClientID,"-"),"[[",1)

saveRDS(y, file = "y_clines.rds")

cs <- colSums(y)
cs <- cs[order(cs)]

par(mar=c(5,10,5,2))

barplot(cs,main="All samples",horiz=TRUE,las=1, xlab="reads")

cols <- ss$DFT
cols <- gsub("DFT1","#B3B9DF",cols)
cols <- gsub("DFT2","#DDAFB7",cols)
mymds <- plotMDS(y,plot=FALSE)

XMIN=min(mymds$x)
XMAX=max(mymds$x)
XMID=(XMAX+XMIN)/2
XMIN <- XMID + (XMIN-XMID)*1.1
XMAX <- XMID+(XMAX-XMID)*1.1
par(mar = c(5.1, 4.1, 4.1, 2.1) )
plotMDS(mymds,pch=19,cex=3,col=cols,main="Cell lines",xlim=c(XMIN,XMAX))
text(mymds,labels=colnames(y))
mtext("blue=DFT1, pink=DFT2")

y_DFT1 <- y[,colnames(y) %in% c("4906_1","4906_2","4906_3",
                                "1426_1","1426_2","1426_3",
                                "C5065_1","C5065_2","C5065_3")]
y_DFT2 <- y[,colnames(y) %in% c("RV_1","RV_2","RV_3",
                                "SN_1","SN_2","SN_3",
                                "TD549_1","TD549_2","TD549_3")]

par(mar = c(5.1, 4.1, 4.1, 2.1) )
mdsDFT1 <- plotMDS(y_DFT1,plot=FALSE)
plotMDS(mdsDFT1,pch=19,cex=3,col="#B3B9DF",main="MDS plot",xlim=c(XMIN,XMAX))
text(mdsDFT1,labels=colnames(y_DFT1))
mtext("DFT1")

par(mar = c(5.1, 4.1, 4.1, 2.1) )
mdsDFT2 <- plotMDS(y_DFT2,plot=FALSE)
plotMDS(mdsDFT2,pch=19,cex=3,col="#DDAFB7",main="MDS plot",xlim=c(XMIN,XMAX))
text(mdsDFT2, labels=colnames(y_DFT2))
mtext("DFT2")

```

## DESeq2 combining the three replicates

Sum replicates. Plot top 40 DEGs.

```{r,de2}

x4906 <- rowSums(y[,ss$clone=="4906"])
xC5065 <- rowSums(y[,ss$clone=="C5065"])
x1426 <- rowSums(y[,ss$clone=="1426"])
xRV <- rowSums(y[,ss$clone=="RV"])
xSN <- rowSums(y[,ss$clone=="SN"])
xTD549 <- rowSums(y[,ss$clone=="TD549"])

y2 <- data.frame(x4906,xC5065,x1426,xRV,xSN,xTD549)
ss2 <- as.data.frame(colnames(y2))
colnames(ss2) <- "clone"
ss2$DFT <- factor(c("DFT1","DFT1","DFT1","DFT2","DFT2","DFT2"))

y2 <- y2[which(rowMeans(y2)>10),]
dim(y2) 

dds <- DESeqDataSetFromMatrix(countData = y2 , colData = ss2, design = ~ DFT )
dds2 <- dds
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds)
zz<-cbind(as.data.frame(z),assay(vsd))
dge<-as.data.frame(zz[order(zz$pvalue),])
dge2 <- dge
sig <- subset(dge,padj<0.05)
sig2 <- sig
sig2_up <- rownames(subset(sig,log2FoldChange>0))
sig2_dn <- rownames(subset(sig,log2FoldChange<0))
length(sig2_up)
length(sig2_dn)

# top 20 up & top 20 down
sig <- sig[order(sig$log2FoldChange, decreasing=TRUE),]

sig_noNA <- sig
sig_noNA$genes <- rownames(sig_noNA)
sig_noNA <- filter(sig_noNA, !grepl(' NA', genes))

top <- rbind(head(sig_noNA,20),tail(sig_noNA,20))

mx <- top[,7:ncol(top)] # get normalised counts

mx.scaled <- t(apply(mx[,1:6], 1, scale)) # center and scale each column (Z-score)
colnames(mx.scaled) <- colnames(mx[,1:6])
colnames(mx.scaled) <- c("4906", "C5065", "1426", "RV", "SN", "TD549")
log2FC <- as.matrix(top$log2FoldChange)
rownames(log2FC) <- rownames(top)
colnames(log2FC) <- "logFC"
mean <- as.matrix(top$baseMean)
rownames(mean) <- rownames(top)
colnames(mean) <- "AveExpr"


col_logFC <- colorRamp2(c(-20,0,20), hcl_palette = "Blue-Red 2")

ha <- HeatmapAnnotation(summary=anno_summary(gp=gpar(fill="#D5D7E2"),height=unit(2, "cm")))

rownames(mx.scaled) <- gsub("^.{0,21}", "", rownames(mx.scaled))
rownames(log2FC) <- gsub("^.{0,21}", "", rownames(log2FC))


top40 <- custom_heatmap(mx.scaled, log2FC, mean, title = "Cell lines")

top40

saveRDS(top40, "top40_clines.RDS")

```

```{r,de2viz}

make_volcano(dge2,"Cell lines")

sig[1:50,1:6] %>%
  kbl(caption="Comparison of DFT1 vs DFT2") %>%
  kable_paper("hover", full_width = F)

write.table(dge2,file="dge2.tsv",sep="\t")

```

## Enrichment analysis

Need to get human homologs. These were obtained from Ensembl v109 using Biomart.

```{r, hom}

rownames(dge2) <- sapply(strsplit(rownames(dge2),"\\."),"[[",1)

hm2 <- hm[hm$Tasmanian.devil.gene.stable.ID != "",]

gt <- hm2[,2:3]
length(unique(gt$Tasmanian.devil.gene.stable.ID))
length(unique(gt$Gene.name))

for(i in 1:length(gt$Gene.name)){
  if(gt$Gene.name[i]==""){gt$Gene.name[i] <- gt$Tasmanian.devil.gene.stable.ID[i]}
}

```

Now run mitch for DGE. The pathways were sourced from Reactome on 7th July 2023. Plot significant enrichments, then top 20, then pathways within Reactome "metabolism" hierarchy.
```{r,mitch1, fig.dim = c(14, 10)}

genesets <- gmt_import("../ref/ReactomePathways_2023-07-14.gmt")

m2 <- mitch_import(dge2, DEtype="deseq2", geneTable=gt)
head(m2)

res2 <- mitch_calc(m2, genesets, priority="effect",cores=16)

if ( !file.exists("mitch_clines_reactome.html") ) {
  mitch_report(res2, "mitch_clines_reactome.html")
}

top <- subset(res2$enrichment_result,p.adjustANOVA<0.05)
saveRDS(top, "~/dftd_RNAseq_annelise/dge/gsea_clines.rds")

top %>%
  kbl(caption="Enriched pathways") %>%
  kable_paper("hover", full_width = F)

ggplot(top, aes(x = s.dist, y = reorder(set, s.dist), color = p.adjustANOVA, size = setSize)) + 
  geom_point(stat = 'identity') + 
  scale_color_viridis(option="mako", direction=1,limits = c(0, 0.05)) +
  xlab("s dist") + ylab("pathway") + ggtitle("Reactome ALL") + 
  theme_bw()

```

```{r,mitch2, fig.dim = c(8, 6)}

top20 <- rbind(slice_max(top, order_by=s.dist, n=10), slice_min(top, order_by=s.dist, n=10))

ggplot(top20, aes(x = s.dist, y = reorder(set, s.dist), color = p.adjustANOVA, size = setSize)) + 
  geom_point(stat = 'identity') + 
  scale_color_viridis(option="mako", direction=1,limits = c(0, 0.05)) +
  xlab("s dist") + ylab("pathway") + ggtitle("Reactome top 20") + 
  theme_bw()

```

```{r,mitch3, fig.dim = c(6, 4)}
react_metabosets <- readRDS("../ref/reactome_metabo.rds")

metabo1 <- filter(top, set %in% names(react_metabosets))
saveRDS(metabo1, "~/dftd_RNAseq_annelise/dge/gsea_clines_metabo.rds")

ggplot(metabo1 , aes(x = s.dist, y = reorder(set, s.dist), color = p.adjustANOVA, size = setSize)) + 
  geom_point(stat = 'identity') + 
  scale_color_viridis(option="mako", direction=1,limits = c(0, 0.05)) +
  xlab("s dist") + ylab("pathway") + ggtitle("Reactome metabolism") + 
  theme_bw()

top_reactome <- top

```

## Visualisation

Heatmaps. First plot enriched pathways, then plot pathways of interest for which experimental data is available.

```{r, complexheatmap1, fig.dim = c(4, 6)}

# see https://www.youtube.com/watch?v=ht1r34-ifVI for reference

hm3 <- data.frame(ENSSHAG=hm2[,2],geneID=hm2[,3])
dge2 <- dge
dge2$ensembl <- rownames(dge2)
split <- strsplit(dge2$ensembl, split = " ")
ensembl <- c()
gene_name <- c()
for (i in 1:length(split)){
  ensembl <- append(ensembl, split[[i]][1]) 
  gene_name <- append(gene_name, split[[i]][2]) 
}
dge2$ensembl <- ensembl
dge2$gene_name <- gene_name

dge3 <- dge2

# only plot significant DEGs
dge3sig <- filter(dge3, padj < 0.05)
dge3sig <- dge3sig[order(dge3sig$log2FoldChange, decreasing=TRUE),]

mx <- dge3sig[,7:ncol(dge3sig)] # get normalised counts
 
mx.scaled <- t(apply(mx[,1:6], 1, scale)) # center and scale each column (Z-score)
colnames(mx.scaled) <- c("DFT1 4906", "DFT1 C5065", "DFT1 1426", "DFT2 RV", "DFT2 SN", "DFT2 TD549")
log2FC <- as.matrix(dge3sig$log2FoldChange)
rownames(log2FC) <- rownames(dge3sig)
colnames(log2FC) <- "logFC"
mean <- as.matrix(dge3sig$baseMean)
rownames(mean) <- rownames(dge3sig)
colnames(mean) <- "AveExpr"

col_logFC <- colorRamp2(c(-20,0,20), hcl_palette = "Blue-Red 2")

custom_heatmap <- function(zscore, fc, aveexpr, title) {
  h1 <- Heatmap(zscore, cluster_rows = F,
              column_labels = colnames(zscore), name="Z-score",
              cluster_columns = T, 
              column_names_gp = gpar(fontsize = 7, fontface="bold"),
             col=colorRamp2(c(-1.5,0,1.5), hcl_palette = "Blue-Red 2"),
             heatmap_legend_param = list(title = "Z-score", at = seq(-1.5, 1.5, 0.5)))
  h2 <- Heatmap(fc, row_labels = rownames(fc), row_names_gp = gpar(fontsize = 6),
              cluster_rows = F, name="logFC", col = col_logFC,column_names_gp = gpar(fontsize =7,fontface="bold"),
              cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
                grid.text(round(as.numeric(fc[i, j],2),2), x, y,gp=gpar(fontsize=6))
              }, column_title = title, 
              heatmap_legend_param = list(title = "logFC", at = seq(-20, 20, 10)))
  
  h <- h1 + h2 
  h
}

ha <- HeatmapAnnotation(summary=anno_summary(gp=gpar(fill="#D5D7E2"),height=unit(1.5, "cm")))

# ----- Top DEGs in REACTOME metabolic pathways ------

metabo <- c() # get all names of genes involved in metabolism
for(i in 1:length(genesets)){metabo <- c(metabo,genesets[[i]])}
metabo <- unique(metabo) # get rid of duplicates

temp <- c() # get ENSSHAG and gene ID
for(i in 1:length(metabo)){temp <- rbind(temp,filter(hm3, geneID==metabo[i]))}
metabolism <- c(t(unite(temp,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

# collapse transcript IDs to gene IDs
rownames(mx.scaled) <- sub('\\..*\\s', ' ', rownames(mx.scaled))
rownames(log2FC) <- sub('\\..*\\s', ' ', rownames(log2FC))
rownames(mean) <- sub('\\..*\\s', ' ', rownames(mean))
dge3$ensembl <- sub('\\..*', '', dge3$ensembl)
rownames(dge3) <- sub('\\..*\\s', ' ', rownames(dge3))

rownames(mx.scaled) <- gsub("^.{0,19}", "", rownames(mx.scaled))
rownames(log2FC) <- gsub("^.{0,19}", "", rownames(log2FC))

metabolism <- gsub("^.{0,19}", "", metabolism)

met <- subset(mx.scaled, rownames(mx.scaled) %in% metabolism)
metFC <- subset(log2FC, rownames(log2FC) %in% metabolism)
metAE <- subset(mean, rownames(mean) %in% metabolism)

# top 20 up & down
met40 <- rbind(head(met,20),tail(met,20))
met40FC <- as.matrix(c(head(metFC,20),tail(metFC,20)))
rownames(met40FC) <- rownames(met40)
colnames(met40FC) <- "logFC"
met40AE <- as.matrix(c(head(metAE,20),tail(metAE,20)))

custom_heatmap(met40, met40FC, met40AE, title = "TOP 40 - REACTOME metabolism")

# ----- All DEGs in REACTOME metabolic pathways ------

dge3_metabo_reactome <- filter(dge3, gene_name %in% metabo)

make_volcano(dge3_metabo_reactome, "Cell lines - Metabolism genes - REACTOME")

# ----- Individual TOP REACTOME metabolic pathways -----

pathways <- metabo1$set
heatmaps <- vector(mode = "list", length = length(pathways))
for(i in 1:length(pathways)){
  
  metabo <- c() # get all names of genes involved in metabolism
  for(j in 1:length(genesets)){metabo <- c(metabo,genesets[[pathways[i]]])}
  metabo <- unique(metabo) # get rid of duplicates

  temp <- c() # get ENSSHAG and gene ID
  for(k in 1:length(metabo)){temp <- rbind(temp,filter(hm3, geneID==metabo[k]))}
  metabolism <- c(t(unite(temp,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))
  
  metabolism <- gsub("^.{0,19}", "", metabolism)
  
  met <- subset(mx.scaled, rownames(mx.scaled) %in% metabolism)
  metFC <- subset(log2FC, rownames(log2FC) %in% metabolism)

  heatmaps[[i]] <- custom_heatmap(met, metFC, title = paste(pathways[i]))
  
}

```

```{r,complexheatmap_1, fig.dim = c(4, 3.25)}
print(heatmaps[[1]])

saveRDS(heatmaps[[1]], "cyst.RDS")
```

```{r,complexheatmap_2, fig.dim = c(4, 3.75)}
print(heatmaps[[2]])

saveRDS(heatmaps[[2]], "chol.RDS")

```

```{r,complexheatmap_3, fig.dim = c(4, 4)}
print(heatmaps[[3]])

saveRDS(heatmaps[[3]], "sulf.RDS")
```

```{r, other, fig.dim = c(4, 5.75)}

# ----- Other interesting pathways REACTOME -----

mm <- filter(hm3, geneID %in% genesets[["Glycolysis"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

glyco_p <- custom_heatmap(met, metFC, title = "Cell lines")

glyco_p

saveRDS(glyco_p, "glyco_p.rds")


```

```{r, other4, fig.dim = c(4, 5.75)}
mm <- filter(hm3, geneID %in% genesets[["Respiratory electron transport, ATP synthesis by chemiosmotic coupling, and heat production by uncoupling proteins."]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

custom_heatmap(met, metFC, metAE, title = "Respiratory electron transport, \n ATP synthesis by chemiosmotic coupling, \n and heat production by uncoupling proteins.")
```
```{r, other5, fig.dim = c(4, 3.25)}

mm <- filter(hm3, geneID %in% genesets[["Citric acid cycle (TCA cycle)"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

custom_heatmap(met, metFC, metAE, title = "Citric acid cycle (TCA cycle)")
```
```{r, other6, fig.dim = c(4, 5.5)}
mm <- filter(hm3, geneID %in% genesets[["Respiratory electron transport"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)
metAE <- subset(mean, rownames(mean) %in% mm)

custom_heatmap(met, metFC, metAE, title = "Respiratory electron transport")
```
```{r, other7, fig.dim = c(4, 4)}
mm <- filter(hm3, geneID %in% genesets[["DNA Repair"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, title = "DNA Repair")
```

```{r, other8, fig.dim = c(4, 6.5)}
mm <- filter(hm3, geneID %in% genesets[["DNA Double-Strand Break Repair"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, title = "DNA Double-Strand Break Repair")
```

```{r, other9, fig.dim = c(4, 4)}
mm <- filter(hm3, geneID %in% genesets[["Detoxification of Reactive Oxygen Species"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

ROS_p <- custom_heatmap(met, metFC, title = "Cell lines")

ROS_p 

saveRDS(ROS_p, "ROS_p.rds")
```
```{r, other99, fig.dim = c(4, 4)}
mm <- filter(hm3, geneID %in% genesets[["Metabolism of nucleotides"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

nucl_p <- custom_heatmap(met, metFC, title = "Metabolism of nucleotides")

nucl_p 

saveRDS(nucl_p, "nucl.rds")
```

```{r, other10, fig.dim = c(4, 3)}
mm <- filter(hm3, geneID %in% genesets[["Pentose phosphate pathway"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, title = "Pentose phosphate pathway")
```

```{r, complexheatmap2, fig.dim = c(4, 8.5)}

mm <- filter(hm3, geneID %in% genesets[["Fatty acid metabolism"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, title = "Fatty acid metabolism")

```

```{r, complexheatmap5,fig.dim = c(4, 2.75)}
mm <- filter(hm3, geneID %in% genesets[["Glutamate and glutamine metabolism"]])
mm <- c(t(unite(mm,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

mm <- gsub("^.{0,19}", "", mm)

met <- subset(mx.scaled, rownames(mx.scaled) %in% mm)
metFC <- subset(log2FC, rownames(log2FC) %in% mm)

custom_heatmap(met, metFC, metAE, title = "Glutamate and glutamine metabolism")
```

## Plotting custom gene sets

Two custom gene sets were added to the previous analysis containing cell competition & metastasis related genes based on a literature search. 

```{r, custom_g, fig.dim = c(4, 3.25)}

cell_compet_genes <- read_excel("../ref/gsea/Cell_compet_genes.xlsx")
names(cell_compet_genes)[names(cell_compet_genes) == "Ensembl_ID"] <- "Gene.stable.ID"

cell_compet_genes <- inner_join(cell_compet_genes,hm,by="Gene.stable.ID") # making sure these genes exist in devils

genesets$"Cell Competition" <- unique(cell_compet_genes$Gene.name)
genesets$"Cell Competition - Mechanical Competition" <- unique(filter(cell_compet_genes, Type == "Mechanical_Competition")$Gene.name)

metastasis_genes <- read_excel("../ref/gsea/Metastasis_genes.xlsx")
names(metastasis_genes)[names(metastasis_genes) == "Ensembl_ID"] <- "Gene.stable.ID"

metastasis_genes <- inner_join(metastasis_genes,hm,by="Gene.stable.ID") # making sure these genes exist in devils

genesets$"Metastasis" <- unique(metastasis_genes$Gene.name)
genesets$"Metastasis - promotors" <- unique(filter(metastasis_genes, Type == "Metastasis_promotor")$Gene.name)
genesets$"Metastasis - suppressors" <- unique(filter(metastasis_genes, Type == "Metastasis_supressor")$Gene.name)

saveRDS(genesets, "~/dftd_RNAseq_annelise/ref/gsea/reactome_plus_custom.RDS")

res_custom <- mitch_calc(m2, genesets, priority="effect",cores=16)

if ( !file.exists("mitch_custom.html") ) {
  mitch_report(res_custom, "mitch_custom.html")
}

cc <- filter(hm3, geneID %in% genesets[["Cell Competition"]])
cc <- c(t(unite(cc,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

cc <- gsub("^.{0,19}", "", cc)

met <- subset(mx.scaled, rownames(mx.scaled) %in% cc)
metFC <- subset(log2FC, rownames(log2FC) %in% cc)

custom_heatmap(met, metFC, title = "Cell competition")
```
```{r, custom_g2, fig.dim = c(4, 5)}
meta <- filter(hm3, geneID %in% genesets[["Metastasis"]])
meta <- c(t(unite(meta,"ENSSHAG_geneID", ENSSHAG:geneID, sep=" ")))

meta <- gsub("^.{0,19}", "", meta)

met <- subset(mx.scaled, rownames(mx.scaled) %in% meta)
metFC <- subset(log2FC, rownames(log2FC) %in% meta)

custom_heatmap(met, metFC, metAE, title = "Metastasis")

```

## Session information

For reproducibility.

```{r,session}

sessionInfo()

```
